content/articles/mocking-with-junit-and-mockito-the-why-and-how/index.md

```diff
@@ -0,0 +1,871 @@
+                                    .password("securedPassword")
+                                    .gender(Gender.MALE).build();
+
+        when(studentRepository.save(any(Student.class))).thenReturn(studentToSave);
+
+        StudentDto studentDto = studentService.registerStudent(registrationRequest);
+
+        assertThat(studentDto, hasProperty("firstName", equalTo("John")));
+        assertThat(studentDto, hasProperty("lastName", equalTo("Doe")));
+        assertThat(studentDto, hasProperty("matricNo", equalTo("MAT100419")));
+        assertThat(studentDto, hasProperty("gender", equalTo(Gender.MALE)));
+    }
+    ```
+
+   The ```save``` method in the ```studentRepository``` class is stubbed to return ```studentToSave``` object whenever we invoke the ```save``` method.
+
+   > The ```any(Student.class)``` [Mockito Argument Matcher](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/ArgumentMatchers.html) means that whenever we invoke the ```save``` method with any object of class ```Student```, the mock ```studentRepository``` will return ```studentToSave``` object.
+
+
+2. ```when- thenThrow(e)```: This throws an exception- ```e``` whenever we invoke the stubbed method. To explain how ```when- thenThrow``` works, let us stub the ```save``` method in our ```StudentRepository```
+    ```java
+    public Student save(Student student) throws HostelManagementException {
+      if (student == null){
+        throw new NullEntityException("student object cannot be null");
+      }
+      database.put(student.getId(), student);
+      return student;
+    }
+    ```
+
+   We can direct our mock ```studentRepository``` to throw a ```NullEntityException``` whenever a null object is passed into the ```save``` method as follows:
+    ```java
+    when(studentRepository.save(null)).thenThrow(NullPointerException.class);
+    ```
+
+3. ```thenAnswer(Answer answer)```: Instead of returning a hard-coded value, we can use the ```thenAnswer(Answer answer)``` stubbing method to return a user-defined value computed at runtime. ```Answer``` is an interface. To demonstrate the usage of the ```thenAnswer()``` stubbing method, let us add a ```registrationTime``` field of type ```LocalDateTime``` to our ```Student``` model.
+
+   ```thenAnswer``` makes sense to use in this scenario because we want our mock to return the exact time a student registered.
+
+   To do this:
+
+   First, we declare a variable ```time``` in our test class and initialize it to ```null``` in our ```@BeforeEach```
+    ```java
+    @ExtendWith(MockitoExtension.class)
+    class StudentServiceImpl_StubbingTests {
+    private StudentService studentService;
+    @Mock
+    private StudentRepository studentRepository;
+    @Mock
+    private HostelRepository hostelRepository;
+    private LocalDateTime time;
+
+    @BeforeEach
+        void setUp() {
+        time = null;
+        studentService = new StudentServiceImpl(studentRepository, hostelRepository);
+      }
+    }
```

Comment:
@ehizman 
The code below the `@BeforeEach` annotation seems to have moved inside abit, could you kindly fix that by replacing it this way :

```java
    @BeforeEach
    void setUp() {
        time = null;
        studentService = new StudentServiceImpl(studentRepository, hostelRepository);
    }
```

Reply:
Done